lib/Sema/CSGen.cpp

```diff
@@ -2552,6 +2552,24 @@ namespace {
             }
           }
 
+          // FIXME: We can see UnresolvedDeclRefExprs here because we have
+          // not yet run preCheckExpression() on the entire closure body
+          // yet.
+          //
+          // We could consider pre-checking more eagerly.
+          if (auto *declRef = dyn_cast<UnresolvedDeclRefExpr>(expr)) {
```

Comment:
Does this mean that this is a body of a multi-statement closure because `preCheckExpression` should walk into single-statement ones and resolve their refs? I think we can move `closure->walk(collectVarRefs);` under `shouldTypeCheckInEnclosingExpression(closure)` to avoid having to handle not-yet-unresolved references, because currently solver handles only single-statement closures...

Reply:
Unfortunately `test/ConstraintSystem/function_builders.swift` ends up with a multi-statement closure here that has not been pre-checked, and if I skip it, the solver fails with unbound free variables.

Reply:
I wonder if pre-check should walk to any closure body then, since it's impossible to say whether it's going to be used as a function builder body?

Reply:
My concern is that since preCheckExpression() uses TypeChecker::lookupUnqualified(), it can end up calling getInterfaceType() on the closure’s parameters for example, so in general it’s not valid to pre-check the body before having applied a solution to the expression containing the multi-statement closure. Avoiding the possibility of getInterfaceType() calls on local parameters with inferred type is also why I implemented lookupSingleLocalDecl() separately.

Reply:
Hm, could you elaborate on why `getInterfaceType()` on a parameter is a problem? Since `preCheckExpression` currently walks into single-statement closures it could find nested closures which wouldn't have any information in the AST from their parents either...

Reply:
I might be misunderstanding but it's concerning because this problem with `getInterfaceType()` on parameter sounds like it would affect any kind of closure...

Reply:
Today, parameter references are resolved earlier in the parser. The unqualified lookup in preCheckExpression() only finds type members and globals, and forward references of local functions.

In general the type checker’s lookup also won’t call getInterfaceType() unless it finds multiple decls and has to do shadowing checks, but that’s not guaranteed.

Reply:
So what does this mean for single-statement closures which don't have their type resolved yet but share parameter names with some other declarations from outer context?

```swift
var x: Int = 0
let _: (Int) -> Void = { x in
  _ = { x, y in x + y }(x, 42)
}
```

here either parameter `x` would be resolved until solution is applied to the AST.

Reply:
It feels like what we really want to do here is instead of mutating AST generate constraints for `UnresolvedDeclRefExpr` directly and that way delay lookup until solver just like we do for members today.